server/src/main/java/org/elasticsearch/action/admin/cluster/node/tasks/cancel/TransportCancelTasksAction.java

```diff
@@ -70,6 +72,7 @@ public TransportCancelTasksAction(ClusterService clusterService, TransportServic
             CancelTasksRequest::new, CancelTasksResponse::new, TaskInfo::new, ThreadPool.Names.MANAGEMENT);
         transportService.registerRequestHandler(BAN_PARENT_ACTION_NAME, ThreadPool.Names.SAME, BanParentTaskRequest::new,
             new BanParentRequestHandler());
+        taskManager.setOrphanedTasksOnChannelCloseListener(this::cancelOrphanedTasks);
```

Comment:
Looks like there is a race condition that leads to an [AssertionError](https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request-1/24072/testReport/junit/org.elasticsearch.search/CrossClusterSearchUnavailableClusterIT/testSearchSkipUnavailable/). See my comment below, I think the logic to cancel a task and its children/descendants should be entirely in the `TaskManager`.

Reply:
Should be fixed in https://github.com/elastic/elasticsearch/pull/56620/commits/bd2950455daec1e115a0cbbdafb986c51bc77585